### PR TITLE
feat(datasets) Experimental - sqlframe datasets

### PR DESCRIPTION
**Pull request by datajoely**
## Description
Leveraging [sqlframe](https://sqlframe.readthedocs.io/en/latest/) a new dataframe library which targets SQL backends (e.g. duckdb/bigquery/postgres) but exposes the PySpark API data frame syntax.... without the JVM or actually running Spark itself. 

This has two major benefits for users:
- Like Ibis it allows users to leverage SQL platforms as an execution engine in addition to a storage engine. Approaches like our `pandas.SQLTableDataset` are naive in the sense they don't use the SQL engine for processing, only storage.
- For users already accustomed to Spark syntax or brownfield projects already written in spark this provides a low-friction adoption route.

## Development notes
- This has been tested locally in the terminal, I've not yet written formal tests. Experimental mode baby 😎 .
- I've also done some funky `OmegaConf` resolver stuff so that the SQL connection can be lazily defined in YAML without creating a super complicated dataset class whilst still supporting dynamic switching of back-ends.

## Checklist

- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the relevant `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

**Review by deepyaman — COMMENTED** [review]
Not a proper review (especially of dataset logic), just left a few quick thoughts.

File: kedro-datasets/kedro_datasets_experimental/sqlframe/resolvers.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    return conn
+
+
+def get_credentials(path) -> str:
+    """Retrieve credentials earlier in the lifecycle"""
+    data = OmegaConf.load("conf/local/credentials.yml")
+    return OmegaConf.select(data, path)
+
+
+from kedro.io.data_catalog import DataCatalog
```
[the remark this answers is not in the document]
Comment (datajoely) [reply kedro-datasets/kedro_datasets_experimental/sqlframe/resolvers.py:34]: Shouldn't have been committed!

File: kedro-datasets/kedro_datasets_experimental/sqlframe/__init__.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+"""Provide data loading and saving functionality for SqlFrame's backends."""
+
+from typing import Any
+
+import lazy_loader as lazy
+
+# https://github.com/pylint-dev/pylint/issues/4300#issuecomment-1043601901
+TableDataset: Any
+FileDataset: Any
```
[the remark this answers is not in the document]
Comment (datajoely) [reply kedro-datasets/kedro_datasets_experimental/sqlframe/__init__.py:9]: Yeah probably now I've worked out how to do it, it's much easier

File: kedro-datasets/kedro_datasets_experimental/sqlframe/resolvers.py
[the remark this answers is not in the document]
Comment (datajoely) [reply kedro-datasets/kedro_datasets_experimental/sqlframe/resolvers.py:1]: So I wanted to discuss this -

Sqlframe doesn't have it's own connector logic, it is built to accept a live connection in the dialect of your choice:

`duckdb`
image

`bigquery`
image

Now the complexity comes in whether we build in declarative and complicated constructor logic for each back-end into the dataset itself. What I'm trying to show here is that OmegaConf resolvers are neat way of us keeping the dataset class dynamic and generic, but I wanted to get the group's view on the topic...

**Comment by noklam** [comment]
What do we want to do about this PR/dataset?

**Comment by datajoely** [comment]
@noklam I'm going to redo it, v2.0.0 just came out and it makes the API much better

**Comment by merelcht** [comment]
@datajoely Are you still keen on polishing this? And maybe a silly question but this looks very similar to the two Ibis datasets we now have - does it make sense to have both?

**Comment by datajoely** [comment]
Hi @merelcht I'm going to close this and reopen this at a later date - the library was changing at a rate of knotty so I held off finishing this.

I do think both this and Ibis deserve to be supported in Kedro - the onboarding penalty in Ibis is the sticking point as you need to change your existing codebase. This is much easier to adopt for existing Spark users and unlocks a bunch of different execution engines. 

My recommendation would be use Ibis if you're starting something new, use SQLFrame if you're thinking about migrating an existing project off Spark.

**Pull request closed without merging**